modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/HeapLockManager.java

```diff
@@ -158,14 +162,22 @@ public HeapLockManager(DeadlockPreventionPolicy deadlockPreventionPolicy, int ma
         }
 
         slots = tmp; // Atomic init.
-
-        parentLockManager.listen(LockEvent.LOCK_CONFLICT, parentLockConflictListener);
     }
 
     @Override
     public CompletableFuture<Lock> acquire(UUID txId, LockKey lockKey, LockMode lockMode) {
-        if (lockKey.contextId() == null) { // Treat this lock as a hierarchy lock.
-            return parentLockManager.acquire(txId, lockKey, lockMode);
+        if (lockKey.contextId() == null) { // Treat this lock as a hierarchy(coarse) lock.
+            // Try fast path.
+            CoarseLockState state = coarseMap.get(lockKey);
+
+            if (state == null) {
+                // Atomically init state.
+                CoarseLockState s = new CoarseLockState(lockKey);
+                CoarseLockState cur = coarseMap.putIfAbsent(lockKey, s);
+                state = (cur == null ? s : cur);
```

Comment:
I think it will be the same:
```
CoarseLockState state = coarseMap.computeIfAbsent(lockKey, key -> new CoarseLockState(lockKey));
```

Reply:
They are not the same.
computeIfAbsent will take a lock on closure computation.
putIfAbsent can work without taking a lock.